ebpf_updates: issue 5 (#114)
cilium/ebpf.io: src/posts/2021-05-03/index.md

  at a $1.19 billion valuation.

<!--2021-04-29-->
* [_Microsoft acquires Kinvolk_](https://kinvolk.io/blog/2021/04/microsoft-acquires-kinvolk/).  
  Launched in 2015, Kinvolk has been working on open source projects such as
  the Flatcar Container Linux distribution, Lokomotive, or Inspektor Gadget.
  The latter is a suite of eBPF-based tools for debugging and inspecting
  Kubernetes clusters, and the company has been actively contributing to the
  eBPF ecosystem (see for example the posts from Mauricio Vásquez Bernal
  referenced later in this document). The company is joining the Azure team,
  congratulations and best of luck for the future!

<!--2021-04-06-->
* [_LPC 2021 Networking and BPF Track CFP_](https://lore.kernel.org/bpf/6d225920-9ecc-ef24-2bf8-848ca86c7fb0@iogearbox.net),
  <span style="white-space: nowrap;">on the development mailing list for eBPF.</span>  
  Just like the previous editions, the next Linux Plumbers Conference will have
  a _Networking and BPF_ track. The call for participation is open until the
  **13th of August**. The conference itself is due to take place in Dublin,

>   Ireland, from the 27th to the 29th of September 2021.

Reviewer: Might want to hold off on the dates here, there's been an update that the conference will go virtual and the dates may need to shift: https://www.linuxplumbersconf.org/blog/2021/index.php/2021/04/30/linux-plumbers-goes-fully-virtual/

Reviewer: Well I assume the CFP won't change (cc @borkmann) but the conf itself might.

Author: Right, thank you! I changed to:
```markdown
  The call for participation is open until the
  **13th of August**. The conference itself was initially due to take place in
  Dublin, Ireland, from the 27th to the 29th of September 2021, but the
  organizers
  [recently announced](...)
  that it will be a fully virtual event instead.
```

Reviewer 2: Yes, LPC is going to be virtual. We'll write that in our next CFP reminder.

  Memcheck, and then turns to perf and eBPF. But the author also strives to
  provide all the technical background to understand what is going on: Stack
  unwinding and how to do it, Linux's memory model, tracing event sources,
  introduction to perf and eBPF, installation steps, are as many sections to
  read and to learn from. At last, a variety of example use cases, often
  involving flame graphs, show how to help troubleshoot memory issues.

<!--2021-04-09-->
* [_Extending systemd Security Features with eBPF_](https://kinvolk.io/blog/2021/04/extending-systemd-security-features-with-ebpf/),
  <span style="white-space: nowrap;">from Mauricio Vásquez Bernal.</span>  
  Two new systemd properties implemented through eBPF programs are in
  development, leveraging the oncoming support for eBPF programs written in C
  (rather than bytecode only) for systemd.

  The first one restricts the file system types that processes in
  a systemd unit can access. Setting `RestrictFileSystems=ext4`, for example,
  will prevent the processes to interact with the tmpfs, thanks to a program
  attached to the eBPF LSM (_Linux Security Module_) and running on all
  attempts to open files.

> 
>   The program checks for the presence of the magic number associated to the
>   file system in a dedicated eBPF map. The second property,

Reviewer 3: ```suggestion
  The program checks for the presence of the magic number associated to the
  file system in a dedicated eBPF map. 
  
  The second property,
```

What about moving the "The program checks for the presence of the magic number associated to the  file system in a dedicated eBPF map" sentence to the previous paragraph?

Author: Right, that was my initial intention, but it looks like I didn't break the line on the correct sentence. Good catch, and thanks for the review!

Updates!

{{/preview}}

Spring is back in the Northern Hemisphere, and with it all kinds of features
are blooming for eBPF. Kernel 5.12 is out, and version 5.13 is on track to
bring new exciting functionalities. At the same time, new projects hatch and
companies are enjoying renewed activity. Several eBPF-related projects applied
to join or upgrade their status in the bosom of CNCF. After some delay, no
doubt caused by winter hibernation, welcome to the fifth issue of the eBPF
Updates!

## Important News

<!--2021-04-12-->
* [Cilium applies to become a CNCF project](https://github.com/cncf/toc/pull/637).  
  The Cilium Community has sent a proposal to move the project into the CNCF
  (_Cloud Native Computing Foundation_) at the Incubation level. If the
  proposal is accepted, there should be very little change to development
  models or governance, already fully open and following open-source best
  practice. From the overview of the proposal, this happens after an increasing
  number of users have requested a donation to a foundation, to get the final
  piece of insurance on the commitment to open-source required for building

>   large projects on top of Cilium. The CNCF was naturally considered, as the

Author: (but what commitment?)

>   large projects on top of Cilium. The CNCF was naturally considered, as the

Reviewer 4: Ah, sorry, I meant to keep "commitment to open source". Fixed.

  be found [on YouTube](https://youtu.be/80OYrzS1dCA). You can suggest topics
  for discussions on the GitHub repository. Make sure you do not miss the next
  episode!

## New Resources

### Introductory Material

<!--2021-03-16-->
* [_eBPF and Kubernetes - What's the Deal?_](https://www.youtube.com/watch?v=ZyT4JB0Ugk8) (video),
  <span style="white-space: nowrap;">from Dawid Ziółkowski.</span>  
  eBPF can greatly improve networking for environment running with Kubernetes.
  After providing some elements of historical context, this presentation
  exposes the limitations imposed to clusters by the use of iptables. It
  introduces eBPF, and then Cilium, and explains how to leverage these tools to
  improve routing, observability, and network policies in clusters.

<!--2021-03-04-->
* [_Getting Started with eBPF and Go_](https://networkop.co.uk/post/2021-03-ebpf-intro/),
  <span style="white-space: nowrap;">from Michael Kashin.</span>  
  Have you tried working with Go and eBPF, and are you getting confused by the
  multiple libraries offering to help manage the programs and the maps?
  Libbpfgo, goebpf, gobpf, cilium/ebpf are several possibilities, wrapping

>   around different components (or in pure Go for the last one). This post aims

Author: `wrappings`?

>   Libbpfgo, goebpf, gobpf, cilium/ebpf are several possibilities, wrapping
>   around different components (or in pure Go for the last one). This post aims

Reviewer 4: Hm, this two clause sentence is maybe correct but definitely weird. Maybe this is better?
```suggestion
  Libbpfgo, goebpf, gobpf, cilium/ebpf are several possibilities. They are
  wrappings around different components or pure-Go implementations
  (for cilium/ebpf). This post aims
```

>   around different components (or in pure Go for the last one). This post aims

Author: Sorry I missed your first comment. No, `wrapping` was not a noun but a verb. We can split, but I'd still keep a verb, `They wrap around`.

>   Libbpfgo, goebpf, gobpf, cilium/ebpf are several possibilities, wrapping
>   around different components (or in pure Go for the last one). This post aims

Reviewer 4: But it doesn't work with "or in pure Go...". Maybe:
```suggestion
  Libbpfgo, goebpf, gobpf, cilium/ebpf are several possibilities. The first
  three wrap around different components, whereas cilium/ebpf has a
  pure Go implementation. This post aims
```

  technology”, like Docker has been.

<!--2021-02-24-->
* [_fledge.io brings eBPF to multi-cloud and edge_](https://www.fledge.io/2021/02/24/fledge-io-brings-ebpf-to-multi-cloud-and-edge/)
  <span style="white-space: nowrap;">from Pramodh Mallipatna.</span>  
  The solution proposes to help define, deploy and manage geo-distributed
  applications that span multi-cloud and edge environments. It now uses eBPF
  programs to monitor the local CPU, disk, and network resources, as well as to
  gather and present richer application information in real time.

<!--2021-02-25-->
* [_Linux Kernel insights with eBPF_](https://community.netdata.cloud/t/linux-kernel-insights-with-ebpf/992?hss_channel=tw-734637019306033152),
  <span style="white-space: nowrap;">from Manos Saratsis.</span>  
  Netdata is already using eBPF to some extent, but has more plans to leverage
  the technology and provide a variety of charts to visualize metrics on all
  components of a system. This post is a roadmap announcement of the features
  to come, to inform the community as well as to gather some feedback. Netdata
  version 1.30.0,
  [released a few weeks later](https://www.netdata.cloud/blog/release-1-30-0/),

>   seems to contain some of these changes.

Author: I hesitated a little, but I guess the recent AWS blog post on Pixie makes it more relevant now.

  This article from LWN.net comments a RFC submitted to
  [the io-uring mailing list](https://lore.kernel.org/io-uring/cover.1613563964.git.asml.silence@gmail.com/t/#u)
  and proposing to add a hook for eBPF programs of a new
  `BPF_PROG_TYPE_IOURING`, to run them in the io\_uring context. With some
  extra efforts, it could lead to making decisions based on the outcome of
  previous operations in the ring. The eBPF program could take decisions on the
  next steps after that operation, such as submitting other I/O operations or
  processing the next file in a list.

<!--2021-04-13-->
* [_Comparing SystemTap and bpftrace_](https://lwn.net/Articles/852112/),
  <span style="white-space: nowrap;">from Emanuele Rocca.</span>  
  SystemTap and bpftrace are tools for dynamically instrument the kernel or
  user applications on Linux, and inspect them in a variety of ways. SystemTap
  is older and compiles its scripts into kernel modules. Bpftrace uses eBPF
  programs attached to different probes. This post from LWN.net exposes the
  differences between the two tools in terms of installation procedure, program
  structure, and features. As one could expect, using kernel modules is

>   ultimately more powerful, but producing eBPF programs with eBPF is faster and

Author: I probably meant to write “with bpftrace”, but ok to change

>   ultimately more powerful, but producing eBPF programs with eBPF is faster and

Author: changed, with `s/easier-to-us/&e/`

  simpler to use.

<!--2021-04-22-->
* [_Toward signed BPF programs_](https://lwn.net/SubscriberLink/853489/2916fbd97c95c700/),
  <span style="white-space: nowrap;">from Jonathan Corbet.</span>  
  Another LWN.net article, this time on a recent proposal sent by Alexei
  Starovoitov in order to introduce support for signed eBPF programs, so that
  the system would only accept authorized eBPF programs. The mechanism is
  compared with the one in place for kernel modules, but signing eBPF programs
  is more complex (due to the usage of maps, to the relocations steps, to the
  bytecode changes that the loader applies before injecting the program). The
  proposal currently under review solves this with a special eBPF program of a

>   new type, which would be allowed to run `bpf()` and `close()` system calls

Author: the type itself does not run anything... :thinking:

>   proposal currently under review solves this with a special eBPF program of a
>   new type, which would be allowed to run `bpf()` and `close()` system calls

Reviewer 4: Ah, right. I read a bit too quickly here. What about:
```suggestion
  proposal currently under review solves this with a new eBPF program type,
  which enables running `bpf()` and `close()` system calls and loading other
  eBPF programs from within the kernel.
```

  and would load other eBPF programs from within the kernel. The work is still
  in progress.

<!--2021-03-11-->
* [_What is vmlinux.h?_](https://www.grant.pizza/blog/vmlinux-header/),
  <span style="white-space: nowrap;">from Grant Seltzer.</span>  
  The vmlinux.h header is generated automatically and contains BTF information
  about the kernel itself. Modern eBPF programs may use this information to
  access to a number of features,
  [CO-RE](https://ebpf.io/blog/ebpf-updates-2020-12#did-you-know-co-re) being
  one of them.

<!--2021-02-18/20-->
* [DevConf.CZ 2021](https://www.devconf.info/cz/) ([Schedule](https://devconfcz2021.sched.com/))

>   hosted several talks related to eBPF. Sadly I could not find the videos _and_
>   slides for all presentations.

Reviewer 4: ```suggestion
  hosted several talks related to eBPF, some of which are summarized below.
  Sadly I could not find the videos _and_ slides for all presentations.
```

    * [_eBPF Iterators_](https://devconfcz2021.sched.com/event/gmK5/ebpf-iterators)
      ([slides](https://static.sched.com/hosted_files/devconfcz2021/60/bpf-iterators.pdf)),
      <span style="white-space: nowrap;">from Jiri Olsa.</span>  
      This is possibly the only document on eBPF iterators we have at the
      moment. Those iterators are hooked on various objects in the kernel (for
      example, among many others: tasks, eBPF programs or maps, active TCP
      connections). Pinned to the eBPF virtual file system, they can be dumped
      (simply with `cat`) to iterate over the selected objects and print a

>       number of information about them. This is very similar, but more

Author: Maybe we should just remove “a number of” in that case? Hard to understand what it does with the info otherwise?

>       number of information about them. This is very similar, but more

Reviewer 4: I'm guessing printing it isn't the only possibility?

>       number of information about them. This is very similar, but more

Author: You process the info and then print the result. I guess you could extract it via a map instead of printing it, for other use cases (although you'd have to `cat` on the pinned path to trigger map updates, which looks like a strange workflow?). But as far as I know, this has been used as an improved version of the procfs so far. I'd like to tell that somehow, and not leave the user wondering “OK but _what do we iterate over the objects for?_”.

>       (simply with `cat`) to iterate over the selected objects and print a
>       number of information about them. This is very similar, but more

Reviewer 4: I get printing is an important use case right now, but surely that's not the only thing you can do with iterators. You can also aggregate information, filter some, etc., no? So what about:
```suggestion
      (simply with `cat`) to iterate over the selected objects and gather
      information to print. This is very similar, but more
```

>       number of information about them. This is very similar, but more

Author: Ah right, you mean it wasn't clear from my phrasing that we can process the info before printing it, is that correct? If so your suggestion looks good, we could use `process` instead of `gather` maybe.

>       (simply with `cat`) to iterate over the selected objects and print a
>       number of information about them. This is very similar, but more

Reviewer 4: Ok, let's go with this then:
```suggestion
      (simply with `cat`) to iterate over the selected objects and process
      information to print. This is very similar, but more
```

      flexible, that what procfs already offers. It is even possible to preload
      and attach some iterators at boot time, so that listing eBPF programs and

>       maps is available at all times. Bpftool, BCC tools, bpftrace, and perf

Author: Does it have to be lowercase? I find it harder to parse the beginning of the sentence. Or maybe swap “BCC tools, bpftrace” and “bpftool”, to solve the issue?

>       maps is available at all times. Bpftool, BCC tools, bpftrace, and perf

Reviewer 4: I agree it's harder to parse, but pretty sure Brendan Gregg would argue to keep it lowercase. That's how it's spelled in his book. Your suggestion makes sense :+1: 
```suggestion
      maps is available at all times. BCC tools, bpftool, bpftrace, and perf
```

>       maps is available at all times. Bpftool, BCC tools, bpftrace, and perf

Author: He can spell it as he pleases, he's not the author of the tool. We could ask Jakub, but honestly... Yep let's just swap the order :) Maybe keep bpftrace close to BCC tools?

(For the record I'm not opposed to keeping tool names lowercase when there is a clear convention, but there are some tools for which it's really unclear to me (bpftool, or libbpf also, for example) and for which I always hesitate when they're at the beginning of a sentence).

>       maps is available at all times. Bpftool, BCC tools, bpftrace, and perf

Reviewer 4: Oops, I thought we were talking about bpftrace somehow :facepalm: I think we had another occurence of this with bpftrace somewhere which I also fixed.

For bpftool I looked at the man page and [that guy's Twitter thread](https://twitter.com/qeole/status/1101452445153222656) and they both keep bpftrace in lowercase, even at the start of sentences :stuck_out_tongue_winking_eye: 

> Maybe keep bpftrace close to BCC tools?

Agreed:
```suggestion
      maps is available at all times. BCC tools, bpftrace, perf, and bpftool
```

>       maps is available at all times. Bpftool, BCC tools, bpftrace, and perf

Author: That guy has no idea what he's talking about. Looks like he mostly avoided the readability issue by _not_ putting “bpftool” at the beginning of sentences (other than beginning of tweets). Anyway.

      all have support for iterators (at various degrees of progress).
    * [_Capturing network traffic in an eXpress Data Path_](https://devconfcz2021.sched.com/event/gmNT/capturing-network-traffic-in-an-express-data-path)
      ([slides](https://static.sched.com/hosted_files/devconfcz2021/ab/Capturing-network-traffic-in-an-eXpress-Data-Path.pdf)),
      <span style="white-space: nowrap;">from Eelco Chaudron.</span>  
      How to inspect network packet with an XDP program attached to the
      interface, when only the `XDP_PASS` return code will pass packets to the
      stack and lead them to a path where tcpdump can see them? To answer this
      question, xdpdump was developed. It relies on the _fentry_ and _fexit_
      hooks that eBPF programs can use to attach to the entry and the exit of
      functions, including other eBPF programs, and inspects the packets before
      and after they are processed by the XDP program. The tool provides a
      command-line interface, and also led to the improvement of a component of

>       the PcapNG capture format. WireShark v3.4.0 and older can benefit from

Author: Not sure the name of said component will speak to many people. I supposed PcapNG would speak more.

From what I understand libpcap != PcapNG (but has limited support for it). I think in this case, the format spec got extended, and Wireshark got improved with some code which is likely shared with xdpdump?

![Screenshot_20210503_233942](https://user-images.githubusercontent.com/17001771/116942181-09bf2f00-ac69-11eb-94c8-a1af39d36520.png)
and
![Screenshot_20210503_234005](https://user-images.githubusercontent.com/17001771/116942182-0a57c580-ac69-11eb-9d8c-f50485c6da03.png)

>       command-line interface, and also led to the improvement of a component of
>       the PcapNG capture format. WireShark v3.4.0 and older can benefit from

Reviewer 4: Ok. Got it now. So what about the following small change?
```suggestion
      command-line interface, and also led to the improvements of the PcapNG
      capture format. WireShark v3.4.0 and older can benefit from
```

>       the PcapNG capture format. WireShark v3.4.0 and older can benefit from

Author: `s/the //` but OK

      on the kind of information that it makes possible to retrieve about the
      kernel. Pahole, a tool used to examine data structure layouts encoded in
      debugging information formats, is central in this talk. Not only do we
      use it to produce BTF information from the kernel, but it also has
      extensive support for BTF and is able to extract information for a number
      of use cases, from pretty-printing type information to extracting
      relevant tokens for kernel live patching.
    * [_Get start with XDP: write a network filter in 10m_](https://devconfcz2021.sched.com/event/gmNo/get-start-with-xdp-write-a-network-filter-in-10m)
      ([video](https://www.youtube.com/watch?v=ZfpAAyfeiwA)),
      <span style="white-space: nowrap;">from Hangbin Liu.</span>  
      This tutorial focuses on XDP and explains how to write a minimal program
      to drop all packets, and to load it with `ip link` or `xdp-loader`. It
      consequently explains how to extend that program to address real use
      cases, to use eBPF maps, and to manage more complex eBPF applications
      with libbpf and libxdp. The sample programs are available

>       [on GitHub](https://github.com/liuhangbin/xdp_examples).

Author: Fair enough, let's drop, but we'll need to change the intro of the Devconf.CZ block to make it clear it's some highlights and not an exhaustive list of eBPF-related talks.

>     * [_Get start with XDP: write a network filter in 10m_](https://devconfcz2021.sched.com/event/gmNo/get-start-with-xdp-write-a-network-filter-in-10m)
>       ([video](https://www.youtube.com/watch?v=ZfpAAyfeiwA)),
>       <span style="white-space: nowrap;">from Hangbin Liu.</span>  
>       This tutorial focuses on XDP and explains how to write a minimal program
>       to drop all packets, and to load it with `ip link` or `xdp-loader`. It
>       consequently explains how to extend that program to address real use
>       cases, to use eBPF maps, and to manage more complex eBPF applications
>       with libbpf and libxdp. The sample programs are available
>       [on GitHub](https://github.com/liuhangbin/xdp_examples).

Reviewer 4: ```suggestion
```

>       [on GitHub](https://github.com/liuhangbin/xdp_examples).

Reviewer 4: Makes sense. I sent a suggestion to change the intro and another to remove that item.

  <span style="white-space: nowrap;">from Filip Busic.</span>  
  This long post goes way beyond what one usually finds in introductions and
  tutorials. It explains how to simply trace memory leaks in applications with
  Memcheck, and then turns to perf and eBPF. But the author also strives to
  provide all the technical background to understand what is going on: Stack
  unwinding and how to do it, Linux's memory model, tracing event sources,
  introduction to perf and eBPF, installation steps, are as many sections to
  read and to learn from. At last, a variety of example use cases, often
  involving flame graphs, show how to help troubleshoot memory issues.

<!--2021-04-09-->
* [_Extending systemd Security Features with eBPF_](https://kinvolk.io/blog/2021/04/extending-systemd-security-features-with-ebpf/),
  <span style="white-space: nowrap;">from Mauricio Vásquez Bernal.</span>  
  Two new systemd properties implemented through eBPF programs are in
  development, leveraging the oncoming support for eBPF programs written in C
  (rather than bytecode only) for systemd.

>   The first one restricts the file system types that processes in a systemd

Author: Let's keep, I've used “file system” everywhere else?

>   The first one restricts the file system types that processes in a systemd

Reviewer 4: Oh, that's right. :+1:

  programs. This does _not_ mean that BCC itself is deprecated: the project is
  simply transitioning to libbpf-based tools and abandoning the Python wrappers
  in newer tracing utilities.

### Academic Works

* [_BMC: Accelerating Memcached using Safe In-kernel Caching and Pre-stack Processing_](https://www.usenix.org/conference/nsdi21/presentation/ghigoff)
  ([PDF](https://www.usenix.org/system/files/nsdi21-ghigoff.pdf)),
  <span style="white-space: nowrap;">from Yoann Ghigoff _et al._</span>  
  Published at NSDI '21 (USENIX), this paper introduces BMC, the _BPF Memcached
  Cache_, a first-level in-kernel cache relying on XDP to accelerate Memcached.
  It improves the response time, by avoiding costly traversals of the network
  stack. The authors provide extensive evaluation of the mechanism, and compare
  it to an improved version of Memcached. Paul Chaignon also wrote
  [a dedicated review of that article](https://pchaigno.github.io/ebpf/2021/04/12/bmc-accelerating-memcached-using-bpf-and-xdp.html),
  rightly concluding that “it's exciting to see BPF applied to application
  protocols”.

### Software Projects

* The Cilium community is proud to

>   [take part in this year's Google Summer of Code](https://summerofcode.withgoogle.com/organizations/4940468447608832/).

Author: Yep, I intended to publish before it was closed :/.

>   [take part in this year's Google Summer of Code](https://summerofcode.withgoogle.com/organizations/4940468447608832/).

Author: Oh but I completely forgot to comment the item at all :open_mouth:. Thanks, Ill use your text.

* The Go library
  [cilium/ebpf v0.5.0](https://github.com/cilium/ebpf/releases/tag/v0.5.0) is
  out, and adds support for attaching programs to kprobes, kretprobes, or
  tracepoints out of the box.

* [ipftrace2 v0.2.0](https://github.com/YutaroHayakawa/ipftrace2/releases/tag/v0.2.0)
  was released and revived the support for tracing the journey of network

>   packets in kernel module functions, thanks to BTF for modules.

Author: Right, let's keep for this time.

### Software Demos and Experiments

* CrowdStrike released [BPFMon](https://github.com/CrowdStrike/bpfmon-example),
  a proof-of-concept utility to map updates in an eBPF map. This is part of a
  study on how to detect attackers who would tamper with some configuration
  options, for example, passed through maps to eBPF programs.

* [Conntracker](https://github.com/rafaeldtinoco/conntracker) is “a firewall
  sniffer”. It analyzes in real time the flow going through Linux's tables
  (netfilter or nf\_tables), and provides output to understand, debug and

>   optimize firewall rules. It can use eBPF to trace TCP and UDP flows and their

Author: :thinking: Not sure on that one??

>   optimize firewall rules. It can use eBPF to trace TCP and UDP flows and their

Reviewer 4: Whoa! Fortunately, you don't blindly accept all my suggested changes :grin:

  to compile USDT (_Userland Statically Defined Tracing_) probes into Rust
  applications, so they can be traced with any tools supporting those probes,
  including for example BCC tools or bpftrace.

* [eBPFSnitch](https://github.com/harporoeder/ebpfsnitch) is an experimental
  application-level firewall for Linux, based on eBPF and NFQUEUE. At this
  time, it filters all outgoing flows, but filtering incoming connections is
  under work. It aims at providing a good integration with containerized
  applications.

* Project [Kube-Knark](https://github.com/chen-keinan/kube-knark) uses the Pcap
  capture format and eBPF programs monitor the calls to the Kubernetes API and
  the changes to the configuration files for the primary node, so as to help
  detect rogue applications that would attempt to change that configuration.

* [QEMU-CSD](https://github.com/Dantali0n/qemu-csd) is “full stack prototype to
  execute BPF programs as if they are running on a Zoned Namespace (ZNS) SSD
  Computational Storage Device (CSD). The entire prototype can be run from
  userspace by utilizing existing technologies such as SPDK and uBPF”, the

>   latter being a user space implementation of an eBPF runtime.

Reviewer 5: @pchaigno sorry for the delayed reply...

FWIW, the link/description seems reasonable to me to keep (as we did). From a quick look, they are testing the idea of using eBPF in the storage device by writing a prototype that implements the NVMe protocol so that it can be used from within qemu as a zoned SSD. Zoned SSDs are SSDs with "smart" partitioning that allows an SSD to be used efficiently from multiple applications.

## The Kernel Side

Kernel 5.12 was released on the 25th of April, and with it a
[number of features](https://kernelnewbies.org/Linux_5.12#Tracing.2C_perf_and_BPF)
discussed in the previous issues made their way into the latest stable version
of Linux.

As for the 5.13 release, the time frame covered by this post covers all its
development cycle. The pull requests below all target that version. As usual,
the description only contains a few highlights. Follow the links to the PRs to

> see the full list of changes.

Author: Makes sense (although splitting by PR is easier for _me_). For future issues I still want some separation between the PR that belong to different dev cycles, though.

LIbbpf can be its own category indeed.

> see the full list of changes.

Author: Batch map operations are not helpers, they're called through `bpf()` from user space. I'll put them into `Misc`.

  <span style="white-space: nowrap;">(Björn Töpel,
  [link](https://lore.kernel.org/bpf/20210308112907.559576-1-bjorn.topel@gmail.com/t/#u))</span>

* Add support for floating point types (float and double) in BTF. The objective
  is to help load programs with BTF information on the s390 architecture.
  <span style="white-space: nowrap;">(Ilya Leoshkevich,
  [link](https://lore.kernel.org/bpf/20210226202256.116518-1-iii@linux.ibm.com/t/#u))</span>

* Document the various sub-commands for the `bpf()` system call (map
  operations, program load, object pinning, and so on). This documentation is
  added to
  [the UAPI header](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/bpf.h?id=f89271f09f589b8e9f98a9d3373d4868d3e668a5#n96),
  just like for the documentation of the eBPF helpers. From there it can be
  parsed and turn into friendlier formats, and should eventually be available
  from
  [the kernel's API guide](https://www.kernel.org/doc/html/latest/driver-api/index.html)
  once version 5.13 is out.
  <span style="white-space: nowrap;">(Joe Stringer,
  [link](https://lore.kernel.org/bpf/20210302171947.2268128-1-joe@cilium.io/t/#u))</span>

> * Extend test runs (`BPF_PROG_TEST_RUN`) to test programs of type

Author: Not sure how familiar people are with the program type. Maybe I could reword this sentence to make it a bit clearer, without the current repetition.

> * Extend test runs (`BPF_PROG_TEST_RUN`) to test programs of type

Reviewer 4: It wouldn't hurt, but honestly if they don't know about this program type, they probably also don't care about the new `BPF_PROG_TEST_RUN` support :slightly_smiling_face:

> * Extend test runs (`BPF_PROG_TEST_RUN`) to test programs of type

Author: Ok let's remove, in particular because the rest of the paragraphs repeats it anyway.

  `BPF_PROG_TYPE_SK_LOOKUP`, to be able to test them and evaluate their

>   performance. This program type is the most recently added at this date, and

Author: “new” sounds like it's being added by the patch set. What about “recently added” or just “recent”?

>   performance. This program type is the most recently added at this date, and

Reviewer 4: Yep, you're right:
```suggestion
  performance. This recent program type
```

  helps selecting a socket for new TCP or UDP flows, and overcomes some of the
  `bind()` limitations in specific use cases.
  <span style="white-space: nowrap;">(Lorenz Bauer,
  [link](https://lore.kernel.org/bpf/20210303101816.36774-1-lmb@cloudflare.com/t/#u))</span>

* Enable task-local storage for tracing programs (only LSM programs could
  access it so far).
  <span style="white-space: nowrap;">(Song Liu,
  [link](https://lore.kernel.org/bpf/20210225234319.336131-1-songliubraving@fb.com/t/#u))</span>

* Add a new helper function `bpf_for_each_map_elem()` to iterate and run a

>   callback function with a given context on all elements of a map. The helper

Author: agreed

* Improve AF\_XDP selftests and program loading. AF\_XDP sockets need a XDP
  program to filter the packets to redirect to user space. But when multiple
  AF\_XDP sockets (“xdpsock” instances) are running on a single interface and
  one of them is terminated, the XDP program would be automatically unloaded,
  thus rendering the other sockets unable to operate. Besides improving the
  selftests, this PR addresses the issue by making libbpf use eBPF “links” to
  properly reference the XDP programs and make them persistent.
  <span style="white-space: nowrap;">(Maciej Fijalkowski,
  [link](https://lore.kernel.org/bpf/20210329224316.17793-1-maciej.fijalkowski@intel.com/t/#u))</span>

* Support calling kernel functions from eBPF programs. This feature has some
  similarities with eBPF helper functions, which are compiled as part of the
  kernel and can be called from eBPF programs. But instead of writing dedicated
  functions, this is about calling pre-existing function from the kernel. This
  does not apply to _any_ function: kernel functions have to be explicitly
  marked as allowed for a given eBPF program type for the call to be permitted

>   by the verifier.

Author: s/BPF/eBPF/ please

>   by the verifier.

Reviewer 4: Fixed. That way pchaigno.github.io will keep ruling "bpf" Google searches :laughing:

  The motivation is to reuse some code portions from the kernel, in particular
  for those eBPF programs that override specific kernel operations
  (`BPF_PROG_TYPE_STRUCT_OPS`), like TCP congestion control.

  A crucial difference with eBPF helpers is that kernel functions that can be
  called are _not_ bounded to a fixed ABI contract. This means that they remain
  free to evolve, even if this breaks existing eBPF programs.

  The eBPF JIT is required for these calls to work. BTF information is also
  necessary.

  Several functions directly related to TCP congestion control are
  [marked as allowed](https://lore.kernel.org/bpf/20210325015124.1543397-1-kafai@fb.com/t/#mbe914ced14b35921b471345b58905f962baf1905)
  for eBPF programs overriding TCP congestion control.
  <span style="white-space: nowrap;">(Martin KaFai Lau,
  [link](https://lore.kernel.org/bpf/20210325015124.1543397-1-kafai@fb.com/t/#u))</span>

* Extend batch map operations (lookup, update, delete) to LPM (_Longest Prefix

>   Match_) maps.

Author: Because patch set uses “LPM” and pretty much all doc on these maps speak of “LPM” and if I were to keep one, “LPM” would feel more natural (to me). I want readers to relate to what they will see elsewhere. But at the same time, I chose to expand for those who are not familiar with the acronym. Do you really think this is an issue?

>   Match_) maps.

Reviewer 4: Ok, makes sense.

  <span style="white-space: nowrap;">(Pedro Tammela,
  [link](https://lore.kernel.org/bpf/20210323025058.315763-1-pctammela@gmail.com/t/#u))</span>

[Fourth one](https://lore.kernel.org/bpf/20210424022459.16039-1-alexei.starovoitov@gmail.com/):

* Extend libbpf's static linker to support extern resolution of global symbols:
  global variables, eBPF sub-programs, and maps defined with BTF information.
  This is a direct follow-up of the patch set included in the second bpf-next
  pull request for the cycle, as described above. Andrii still intends to
  address a few follow-up issues, but this set already makes the linker fully
  able to combine variables, maps and sub-programs from several object files
  into a single eBPF program to load into the kernel.
  <span style="white-space: nowrap;">(Andrii Nakryiko,

>   [link](https://lore.kernel.org/bpf/20210423181348.1801389-1-andrii@kernel.org/t/#u))</span>

Author: Ok why not, as long as we keep the two links
